src/main/java/net/logstash/logback/appender/AbstractBeatsTcpSocketAppender.java

```diff
@@ -0,0 +1,373 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.logstash.logback.appender;
+
+import ch.qos.logback.core.encoder.Encoder;
+import ch.qos.logback.core.encoder.EncoderBase;
+import ch.qos.logback.core.spi.DeferredProcessingAware;
+import net.logstash.logback.appender.listener.TcpAppenderListener;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.Socket;
+import java.net.SocketTimeoutException;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static net.logstash.logback.util.ByteUtil.bytesToInt;
+import static net.logstash.logback.util.ByteUtil.intToBytes;
+
+/**
+ * An appender that is compatible with Lumberjack v2 protocol (Beats input).
+ *
+ * Implemented according to current <a href="https://github.com/logstash-plugins/logstash-input-beats/blob/master/PROTOCOL.md">Lumberjack specification</a>.
+ * Verified with <a href="https://github.com/Graylog2/graylog2-server/blob/f35df42/graylog2-server/src/main/java/org/graylog/plugins/beats/Beats2Codec.java">Graylog's Lumberjack v2 implementation</a>.
+ */
+public abstract class AbstractBeatsTcpSocketAppender<Event extends DeferredProcessingAware, Listener extends TcpAppenderListener<Event>>
+        extends AbstractLogstashTcpSocketAppender<Event, Listener> {
+
+    private static final byte PROTOCOL_VERSION = '2';
+
+    private static final byte PAYLOAD_JSON_TYPE = 'J';
+    private static final byte PAYLOAD_WINDOW_TYPE = 'W';
+    private static final byte PAYLOAD_ACK_TYPE = 'A';
+
+    /**
+     * Counter for sequence number of published events.
+     * Sent to remote Beats input to track how many messages were sent
+     * and used to track when to wait for remote to be ready to accept new logs.
+     */
+    private final AtomicInteger counter = new AtomicInteger();
+    /**
+     * A queue of currently accepted ACKs from remote
+     */
+    private final BlockingQueue<AckEvent> ackEvents = new ArrayBlockingQueue<>(10);
```

Comment:
Do we need to keep a queue of the last 10 received ACKs or is it enough to keep the latest only? 
AFAIK an ACK received for sequence `x` acknowledge all events _up to_ `x` included.